src/org/opendatakit/briefcase/export/CsvMapper.java

```diff
@@ -0,0 +1,334 @@
+
+    String binaryFilename = element.getValue();
+
+    if (!exportMedia)
+      return singletonList(Pair.of(element.fqn(), binaryFilename));
+
+    if (!Files.exists(exportMediaPath))
+      createDirectories(exportMediaPath);
+
+    int dotIndex = binaryFilename.lastIndexOf(".");
+    String namePart = (dotIndex == -1) ? binaryFilename : binaryFilename.substring(0,
+        dotIndex);
+    String extPart = (dotIndex == -1) ? "" : binaryFilename.substring(dotIndex);
+
+    File binaryFile = new File(workingDir.toFile(), binaryFilename);
+    int version = 1;
+    File destFile = exportMediaPath.resolve(binaryFilename).toFile();
+    boolean exists = false;
+    String binaryFileHash;
+    String destFileHash;
+
+    // TODO Refactor this and add semantics
+    if (destFile.exists() && binaryFile.exists()) {
+      binaryFileHash = getMd5Hash(binaryFile.toPath());
+
+      while (destFile.exists()) {
+        if (fileHashMap.containsKey(destFile.getName())) {
+          destFileHash = fileHashMap.get(destFile.getName());
+        } else {
+          destFileHash = getMd5Hash(destFile.toPath());
+          if (destFileHash != null) {
+            fileHashMap.put(destFile.getName(), destFileHash);
+          }
+        }
+
+        if (destFileHash != null && destFileHash.equals(binaryFileHash)) {
+          exists = true;
+          break;
+        }
+
+        String otherDestBinaryFilename = namePart + "-" + (++version) + extPart;
+        destFile = exportMediaPath.resolve(otherDestBinaryFilename).toFile();
+      }
+    }
+    if (binaryFile.exists() && !exists) {
+      copy(binaryFile.toPath(), destFile.toPath());
+    }
+    return singletonList(Pair.of(element.fqn(), "media" + File.separator + destFile.getName()));
+  }
+
+  private static List<Pair<String, String>> repeatableGroup(String localId, Model current, XmlElement element) {
+    int shift = current.countAncestors() - 1;
+    return element == null
+        ? empty("SET-OF-" + current.fqn(shift))
+        : singletonList(Pair.of(current.fqn(), localId + "/" + current.fqn(shift)));
+  }
+
+  private static List<Pair<String, String>> nonRepeatableGroup(String localId, Path workingDir, Model current, Optional<XmlElement> maybeElement, boolean exportMedia, Path exportMediaPath) {
+    return current.flatMap(field -> getMapper(field).apply(
+        localId,
+        workingDir,
+        field,
+        maybeElement.flatMap(element -> element.findElement(field.getName())),
+        exportMedia,
+        exportMediaPath
+    ));
+  }
+
+  @SuppressWarnings("checkstyle:OverloadMethodsDeclarationOrder")
+  private static List<Pair<String, String>> empty(String fqn) {
+    return empty(fqn, 1);
+  }
+
+  private static List<Pair<String, String>> formattedDate(XmlElement element, DateFormat formatter) {
+    String value = element.getValue();
+    return singletonList(Pair.of(element.fqn(), value.isEmpty() ? null : formatter.format(parseDate(value))));
+  }
```

Comment:
This fails when the date tag has no body. Here’s a guess about fixing it, but maybe you have some better chaining construct, or a different solution altogether.
![screen shot 2018-05-12 at 10 01 07 pm](https://user-images.githubusercontent.com/55099/39964047-3c2c0086-5630-11e8-9d87-2d217936ca3b.png)

Reply:
Oh, I see for binary you do this:
```
    if (!element.hasValue())
      return empty(element.fqn());
```

Reply:
I see that it's not consistent how I deal with empty elements through the code at this point... Let's discuss how it should behave when dealing with empty elements and then I'll change this commit.

We need to discuss these data types:

| Data type | Expected output |
| --- | --- |
| text (generic field) | an empty string |
| date & time types | 1 empty string |
| geopoint | 4 empty strings |
| binary | 1 empty string |
| null (non repeating group) | should throw? |
| null (repeating group) | should throw? |

Reply:
I don’t know. Can you just do what the previous export did?